draft-yasskin-dispatch-bundled-exchanges.md

```diff
@@ -0,0 +1,675 @@
+1. If `bundleLength > bytes.length`, return an error.
+1. Let `stream` be a new stream whose:
+   * Available bytes are `[bytes[bytes.length - bundleLength],
+     bytes[bytes.length])`.
+   * EOF flag is set.
+   * Current offset is initially 0.
+   * The wait for N bytes operation succeeds immediately if `currentOffset + N
+     <= bundleLength` and fails otherwise.
+1. Return the result of running {{load-metadata}} with `stream` as input.
+
+## Load a response from a bundle {#load-response}
+
+The result of {{load-metadata}}{:format="title"} maps each request to a
+response, which consists of headers and a payload. The headers can be loaded
+from the bundle's stream before waiting for the payload, and similarly the
+payload can be streamed to downstream consumers.
+
+~~~~~ cddl
+response = [headers: bstr .cbor headers, payload: bstr]
+~~~~~
+
+To implement {{semantics-load-response}}, the parser MUST run the following
+steps, taking the bundle's `stream` and one `request` and its `requestMetadata`
+as returned by {{semantics-load-metadata}}.
+
+1. Seek to the start of `stream`, and wait for `requestMetadata.offset + 10`
+   bytes (array item header + the maximum length of a byte string item header)
+   to be available. If this fails, return an error.
+1. Seek forward by `requestMetadata.offset` bytes.
+1. Read 1 byte from `stream`. If it isn't `0x82`, return an error.
+1. Let `headerLength` be the result of getting the length of a CBOR bytestring
+   header from `stream` ({{parse-bytestring}}). If `headerLength` is an error,
+   return that error.
+1. If `headerLength` is TBD or greater, return an error.
+1. Wait for `headerLength` bytes to be available on `stream`. If this fails,
+   return an error.
+1. Let `headerCbor` be the result of reading `headerLength` bytes from `stream`
+   and  parsing a CBOR item from them matching the `headers` CDDL rule. If this
+   returns an error, return that error.
+1. Let `headers`/`pseudos` be the result of converting `cbor-http-request` to a
+   header list and pseudoheaders using the algorithm in {{cbor-headers}}. If
+   this returns an error, return that error.
+1. If `pseudos` does not have a key named ':status' or its size isn't 1, return
+   an error.
+1. If `pseudos[':status']` isn't exactly 3 ASCII decimal digits, return an
+   error.
+
+1. Let `payloadLength` be the result of getting the length of a CBOR bytestring
+   header from `stream` ({{parse-bytestring}}). If `payloadLength` is an error,
+   return that error.
+1. If `stream.currentOffset + payloadLength != requestMetadata.offset +
+   requestMetadata.length`, return an error.
+
+1. Let `body` be a new body ({{FETCH}}) whose stream is a tee'd copy of `stream`
+   starting at the current offset and ending after `payloadLength` bytes.
+
+   TODO: Add the rest of the details of creating a `ReadableStream` object.
+
+1. Let `response` be a new response ({{FETCH}}) whose:
+   * Url list is `request`'s url list,
```

Comment:
"Parsing the index section" seems to say request has one url, do you mean it could have a list / redirect chain?

Reply:
This is just an asymmetry in how requests and responses default their values: https://fetch.spec.whatwg.org/#concept-request-url-list defaults from the request's url, while https://fetch.spec.whatwg.org/#concept-response-url is hard-coded to point to the response's url list. I believe both URL lists should always have 1 value for exchanges in the bundle.

I suspect Fetch will actually wind up copying or extending this response when serving browser-generated requests from bundles, so that a redirect into a bundle can work correctly.